initial draft of vtctldserver blog post (#1417)
vitessio/website: content/en/blog/2023-03-08-vtctldserver-api.md

---
title: "Vtctld Server API"

> date: 2023-03-08T06:30:10-05:00

Reviewer: Can you change this to Monday's date? And leave out the time part of it, have just the date.
Also, can you change the file name to also be Monday's date?

Author: yep! i'll also make [this change](https://github.com/vitessio/website/pull/1417#discussion_r1167181542) while i'm doing that

---
title: "Vtctld Server API"
date: 2023-03-08T06:30:10-05:00
slug: '2023-03-08-vtctldserver-api'
tags:
- Vitess
- vtctld
- Cluster Management
description: A structured gRPC API for Vitess cluster management.
---

> We are more than thrilled to belatedly discuss `VtctldServer`, the new gRPC API for performing cluster management operations with `vtctld` components, which became generally available in Vitess v15.

Author: @deepthi wdyt about this one (imo it's better in some ways - clarifies the versions, but worse in others - kinda wordy and nearing on run-on)

Reviewer: I'd cut a few words, left a suggestion.
In addition to cutting those words, you could also move the "GA" part of it to the next sentence. Like so:
```
began back in [Vitess v9] and went GA in Vitess v15
```

> We are more than thrilled to belatedly discuss `VtctldServer`, the new gRPC API for performing cluster management operations with `vtctld` components, which became generally available in Vitess v15.

Reviewer: ```suggestion
We are thrilled to discuss `VtctldServer`, the new gRPC API for performing cluster management operations with `vtctld` components, which became generally available in Vitess v15.
```

    // ... logic goes here
}
```
<br/>

becomes the following RPC:

```proto
// proto/vtctlservice.proto
service Vtctld {
    rpc ApplyRoutingRules(vtctldata.ApplyRoutingRulesRequest) returns (vtctldata.ApplyRoutingRulesResponse) {};
}

// ===============

// proto/vtctldata.proto
message ApplyRoutingRulesRequest {
    vschema.ShardRoutingRules shard_routing_rules = 1;
    bool skip_rebuild = 2;
    repeated string rebuild_cells = 3;
}

message ApplyRoutingRulesResponse {
}
```
<br/>

This structure allows us to provide a well-defined interface for each individual cluster management action, as well as understand if a change is breaking compatibility between versions, neither of which were possible with the old structure.

Note that the `--dry-run` and `--rules_file` options are handled on the client side, and so did not make it into the request message definition. See &mdash; exceptions!.

There are few other general exceptions[^1] to call out, so you know what to look for when searching for a particular command or RPC.

#### Exception 1: Consolidation

> First, in the old model, occasionally there were several instances of similar commands that we have compressed into a single RPC with different options to switch between the subtle behavioral differences.

Reviewer: ```suggestion
In the old model, there were several instances of similar commands that we have compressed into a single RPC with different options to switch between the subtle behavioral differences.
```

    repeated string cells = 3;
    // Strict specifies how the server should treat failures from individual
    // cells.
    //
    // When false (the default), GetTablets will return data from any cells that
    // return successfully, but will fail the request if all cells fail. When
    // true, any individual cell can fail the full request.
    bool strict = 4;
    // TabletAliases is an optional list of tablet aliases to fetch Tablet objects
    // for. If specified, Keyspace, Shard, and Cells are ignored, and tablets are
    // looked up by their respective aliases' Cells directly.
    repeated topodata.TabletAlias tablet_aliases = 5;
    // tablet_type specifies the type of tablets to return. Omit to return all
    // tablet types.
    topodata.TabletType tablet_type = 6;
}

message GetTabletsResponse {
    repeated topodata.Tablet tablets = 1;
}
```
<br/>

So, depending on which fields you set in a `GetTablets` call, you will get either the behavior of `ListTablets`, `ListAllTablets`, or `ListShardTablets`.
Meanwhile, the `GetTablet` RPC is a 1-to-1 drop-in for the legacy `GetTablet` command.

#### Exception 2: Pluralization

> Second, certain commands would only operate on a single instance of a resource at a time.

Reviewer: ```suggestion
Certain commands would only operate on a single instance of a resource at a time.
```

For example, if you wanted to delete N shards, you needed to make N round-trips to a `vtctl` by invoking N `DeleteShard` commands.

> For these sorts of commands, we've tried to "pluralize" them, to operate on multiple reserouces with a single round-trip to the `vtctld`.

Reviewer: ```suggestion
For these sorts of commands, we've tried to "pluralize" them, to operate on multiple resources with a single round-trip to the `vtctld`.
```

So, `DeleteShard` becomes `DeleteShards`, and `DeleteTablet` becomes `DeleteTablets`, and so on.

> For the destructive operations (like the two `Delete*` RPCs above), we perform them sequentially, and, if any instance fails (i.e. we fail to delete the 3rd tablet), the overall RPC returns an error.

Reviewer: ```suggestion
For destructive operations (like the two `Delete*` RPCs above), we perform them sequentially, and, if any instance fails (i.e. we fail to delete the 3rd tablet), the overall RPC returns an error.
```

#### Exception 3: Streaming

As we said earlier, the old gRPC API had a single, _streaming_ RPC, through which all `vtctl` commands were proxied.

> In most &mdash; but not all! &mdash; cases, these "streamed" responses only ever consisted of one packet from the server, but to properly consume stream, you needed to write a receive loop that would only ever iterate once:

Reviewer: ```suggestion
In most &mdash; but not all! &mdash; cases, these "streamed" responses only ever consisted of one packet from the server, but to properly consume the stream, you needed to write a receive loop that would only ever iterate once:
```

```go
stream, err := client.ExecuteVtctlCommand(
    ctx,
    []string{"GetTablet", "zone1-101"},
    24 * time.Hour,
)
if err != nil {
    // Fail.
}
defer stream.Close()

for {
    e, err := stream.Recv()
    switch err {
    case nil:
        // Marshal the event's bytes into a tablet structure
        // and carry on with what you actually cared about doing.
    case io.EOF:
        break
    default:
        // Fail.
    }
}
```
<br/>

Silly!
That's why we moved to unary RPCs, which allows you to write the above code with the new API as:

```go
resp, err := client.GetTablet(
    ctx,
    &vtctldatapb.GetTabletRequest{
        Alias: &topodatapb.TabletAlias{Cell: "zone1", Uid: 101},
    },
)
if err != nil {
    // Fail.
}

// Do something with resp.Tablet, which is already a
// strucutred Tablet object.
```
<br/>

Ahhhh, isn't unary so much cleaner?

However, there are a few cluster management operations that legitimately benefit from a streaming model between the client and the server.
In those few cases, where we may have an extremely long-running operation, or want some sort of progress indication, we've kept the streaming response paradigm.

> `Backup` and `Restore` are the two most clear examples, but again, we recommend you check against the actual RPC definition at time of programming for the authoritative source.

Reviewer: ```suggestion
`Backup` and `Restore` are the two clearest examples, but again, we recommend you check against the actual RPC definition at time of programming for the authoritative source.
```

Reviewer: Or `two obvious examples` also sounds fine.

Author: i want to avoid the word "obvious" if at all possible

> <!-- aside about how state machine commands like `Reshard` are actually better for these systems? -->

Reviewer: Delete this comment now?

Author: yep!

### Errors

In addition to the "one RPC per command" remit of the new API, the other noteworthy element of our gRPC implementation of that API is a revisiting of errors.
The old API's implementation almost exclusively[^2] returned plain Go `error` types back up to the `ExecuteVtctlCommand` implementation, which were dutifully translated by `grpc-go` into `UNKNOWN` errors, which is ... not super helpful.

> When implementing the new API, we tried to, wherever possible, use the `vterrors` package to surface more helpful information back to the caller wherever possible.

Reviewer: ```suggestion
When implementing the new API, we tried to, wherever possible, use the `vterrors` package to surface more helpful information back to the caller.
```

Author: lol whoops

---
title: "Vtctld Server API"
date: 2023-03-08T06:30:10-05:00
slug: '2023-03-08-vtctldserver-api'
tags:
- Vitess
- vtctld
- Cluster Management
description: A structured gRPC API for Vitess cluster management.
draft: true
---

> We are more than thrilled to announce the general availability of `VtctldServer`, a new gRPC API to perform cluster management operations with `vtctld` components in your Vitess clusters.

Reviewer 2: We should include the version of Vitess in which we are marking it GA

Reviewer: Didn't it go GA in v15?
This should be rewritten to reflect that.

Author: Yeah v15, but we didn't talk about it (enough, imo) until now. Let me fiddle with this wording a bit

This is the (near-) culmination of long, steady migration that began back in [Vitess v9][initial_pr] (!!!), so we'd like to talk a bit about the motivation behind the move, the design of the new API, and where we go from here.

## Why?

> Vitess users may have found themselves invoking some various cluster management operations (think `CreateKeyspace`, `EmergencyReparentShard`, `Backup`, and so on) via the `vtctlclient` program.

Reviewer: ```suggestion
Vitess users may have found themselves invoking various cluster management operations (think `CreateKeyspace`, `EmergencyReparentShard`, `Backup`, and so on) via the `vtctlclient` program.
```

These commands communicate with a remote `vtctld` in the cluster via a gRPC interface defined as follows:

```proto
// proto/vtctlservice.proto

service VtctlServer {
    rpc ExecuteVtctlCommand(ExecuteVtctlCommandRequest) returns (stream ExecuteVtctlCommandResponse) {};
}

message message ExecuteVtctlCommandRequest {
    repeated string args = 1;
    int64 action_timeout = 2;
}

message ExecuteVtctlCommandResponse {
    Event event = 1;
}

// ===============

// proto/logutil.proto
message Event {
    // other fields omitted
    string value = 5;
}
```
<br/>

As the [RFC] points out, there are several issues with this design.

> First, and most noticable, this is effectively an untyped interface, precisely the opposite of one of the biggest benefits of using an interface definition language (IDL) like gRPC.

Reviewer: notable? vs noticeable?
I'm ok with either, if the latter, the typo should be fixed.

Author: ahaha i meant to write "noticeably" so, neither! :P

Second, and much more subtly, this single RPC is a _streaming_ method.
This is necessary to support commmands such as `Backup`, which need to run uninterrupted by a closed connection, potentially for hours.
However, the design of `vtctld`'s logger essentially `tee`'s log lines both to a file and to the client.
This is used to send responses to the client _and_ to log useful information to a cluster operation for later inspection and debugging.
This means that _even if_ (1) you know the structure of the response data and (2) we're careful to never break that between versions, we can _still_ break your response parsing if we add a log line to a given command.
Computers are hard!

## `VtctldServer`

To solve these problems, we decided to [introduce a new gRPC interface to the `vtctld`][initial_pr], with the intention of replacing and eventually retiring the old interface.

It's defined in the same protobuf file, which results in the (possibly confusing) import name of [`vtctlservice.VtctldServer`][vtctldserver_protobuf].

### Structure

> (A quick preface: We're going to mostly speak in generalities here, because to detail each individual special case, exception, and difference could turn this blog post into a novel-length [reamde][reamde].

Reviewer 2: I like the reamde reference

> (A quick preface: We're going to mostly speak in generalities here, because to detail each individual special case, exception, and difference could turn this blog post into a novel-length [reamde][reamde].

Reviewer: I don't think we need the parens around this block, but that's a stylistic choice that I'll leave up to you.

Author: removed!

	routingRulesFile := subFlags.String("rules_file", "", "Specify rules in a file")
	skipRebuild := subFlags.Bool("skip_rebuild", false, "If set, do no rebuild the SrvSchema objects.")
	dryRun := subFlags.Bool("dry-run", false, "Do not upload the routing rules, but print what actions would be taken")
	var cells []string
	subFlags.StringSliceVar(&cells, "cells", cells, "If specified, limits the rebuild to the cells, after upload. Ignored if skipRebuild is set.")

    // ... logic goes here
}
```
<br/>

becomes the following RPC:

```proto
// proto/vtctlservice.proto
service Vtctld {
    rpc ApplyRoutingRules(vtctldata.ApplyRoutingRulesRequest) returns (vtctldata.ApplyRoutingRulesResponse) {};
}

// ===============

// proto/vtctldata.proto
message ApplyRoutingRulesRequest {
    vschema.ShardRoutingRules shard_routing_rules = 1;
    bool skip_rebuild = 2;
    repeated string rebuild_cells = 3;
}

message ApplyRoutingRulesResponse {
}
```
<br/>

This structure allows us to provide a well-defined interface for each individual cluster management action, as well as understand if a change is breaking compatibility between versions, neither of which were possible with the old structure.

> Note that the `--dry-run` and `--rules_file` options are handled on the client side, and so did not make it into the request message definition. See &mdash; exceptions!.

Reviewer 2: I don't get why we have a long dash before exceptions.

Author: Colon would also work here; comma, less so

Reviewer: I get what Manan is saying, punctuation can be left out altogether, like
```
See exceptions.
```
WDYT?

Author: that phrasing removes the tone i was going for (jovial, "computers are hard haha!!") (also that period i had here is a mistake and should be removed, just the `!`)

Reviewer: Sure, let's keep it then.

    }
}
```
<br/>

Silly!
That's why we moved to unary RPCs, which allows you to write the above code with the new API as:

```go
resp, err := client.GetTablet(
    ctx,
    &vtctldatapb.GetTabletRequest{
        Alias: &topodatapb.TabletAlias{Cell: "zone1", Uid: 101},
    },
)
if err != nil {
    // Fail.
}

// Do something with resp.Tablet, which is already a
// strucutred Tablet object.
```
<br/>

Ahhhh, isn't unary so much cleaner?

However, there are a few cluster management operations that legitimately benefit from a streaming model between the client and the server.
In those few cases, where we may have an extremely long-running operation, or want some sort of progress indication, we've kept the streaming response paradigm.
`Backup` and `Restore` are the two most clear examples, but again, we recommend you check against the actual RPC definition at time of programming for the authoritative source.

<!-- aside about how state machine commands like `Reshard` are actually better for these systems? -->

### Errors

In addition to the "one RPC per command" remit of the new API, the other noteworthy element of our gRPC implementation of that API is a revisiting of errors.

> The old APIs implementation almost exclusively[^2] returned plain Go `error` types back up to the `ExecuteVtctlCommand` implementation, which were dutifully translated by `grpc-go` into `UNKNOWN` errors, which is ... not super helpful.

Reviewer: ```suggestion
The old API's implementation almost exclusively[^2] returned plain Go `error` types back up to the `ExecuteVtctlCommand` implementation, which were dutifully translated by `grpc-go` into `UNKNOWN` errors, which is ... not super helpful.
```

It's also worth noting that many of the old API commands have been refactored to use the implementation powering the new API under the hood, with a small translation layer to transform the responses back into their old data structures ([like these](https://github.com/vitessio/vitess/blob/8f68f3f72fe8151beb531e7a16eef1abb0314809/go/vt/vtctl/vtctl.go#L1439-L1526)), if you needed some additional confidence in the correctness of the new implementation.

## Example: VTAdmin

The primary consumer of the new API within the Vitess codebase is [VTAdmin][vtadmin_blog_post].
It uses the structured gRPC interface to perform cluster management operations on all of the clusters it's configured to manage.
If you're looking for an example of the API from a client-side perspective, this is a good place to look.

A fair warning, though &mdash; there's a small bit of indirection to allow VTAdmin to proxy through "some `vtctld` in the cluster" as opposed to needing to dial a _particular_ `vtctld`, but from a "how do I use this API" perspective, just look for `cluster.Vtctld.{RPCMethodName}(...)` calls inside `go/vt/vtadmin`.

## Example: Audit backups

> The other big benefit to the structured API is it's (relatively) easy to program against.

Reviewer: ```suggestion
The other big benefit to the structured API is that it's (relatively) easy to program against.
```

                log.Printf(err)
            }
        }(ksShard[0], ksShard[1])
    }

    wg.Wait()
}

```
</br>

Neat, huh?

## `vtctldclient`

Of course, we're not going to make you write a bunch of code if you just want to invoke an RPC or two.
Similar to the old model, which included a `vtctlclient` binary for invoking vtctl commands on a remote server, we now include a `vtctldclient` for making RPCs to a remote server using the new API.

So, if you just need to fire off a quick RPC, or are feeling particularly bold and want to write the above example in a shell script, you can use the new binary for it:

```
$ vtctldclient --server ":15999" GetBackups "ks/-"
$ vtctldclient --server ":15999" ApplySchema --sql "CREATE TABLE foo (id INT(11) NOT NULL)" "my_other_ks"
```
<br/>

For more information, you can check the [reference docs][vtctldclient_reference_docs], or run `vtctldclient help` or `vtctldclient help <command>`.

A quick but important callout &mdash; note the additional "d"!
It's subtle, and the sheer number of consecutive, full-height consonants does not help the matter, but we've found it's very easy for old habits to omit it.

> Omitting it, of course, results in your uninentionally using the old binary, which isn't what you wanted, probably won't work in subtle ways, and soon won't work at all.

Reviewer: ```suggestion
Omitting it, of course, results in your unintentionally using the old binary, which isn't what you wanted, probably won't work in subtle ways, and soon won't work at all.
```
